### PR TITLE
121 implement authcontext for user authentication in react with typescript

### LINKED ISSUE #121: Implement AuthContext for User Authentication in React with TypeScript
The **AuthContext** is essential for managing user authentication in a centralized manner

**Tasks:**

- Set up a new AuthContext component in the application.
- Define the initial state for user authentication.
- Implement TypeScript for static typing within the AuthContext.
- Define TypeScript interfaces for authentication-related data structures.
- Implement functions within the AuthContext for user login and logout.
- Include functions for checking the authentication status.
- Develop a Context Provider that wraps the application, providing authentication state and functions to child components.
- Integrate the AuthContext with the Navbar component to display different navigation options based on user authentication status.
- Write unit tests for the AuthContext functions using Jest and React Testing Library.
- Conduct thorough testing to ensure reliable authentication behaviour.
- Implement error handling within the AuthContext, providing clear feedback for authentication-related issues.

**Note:**

- As usual please ensure extensive use of comments, Jest tests, error handling and validation where applicable.
- The estimate for this is 4 days

### PR DESCRIPTION
**Pull request by glenntanjoh**
Resolves #121  (issue)

# Description

Added **`LoginForm.tsx`**, **`AuthContext.tsx`**, **`PrivateRoute.tsx`**, **`RequireAuth.tsx`**, **`UserContext.tsx`**, **`Api.ts`**, **`Api.test.tsx`**. Also installed **`jest-mock-fetch`** to test login and logout functions.
Also edited app.tsx by adding RequireAuth.tsx to contain protected pages.

**_For this task, I created a new user directly in the database using Adminer. I planned on using it for all the authentication needs_**
The logic aspect of it(Doesn't allow you to access pages till logged in) works but I can't seem to log in using the user I created and test if it does this correctly.

What each file does(or is supposed to do):

1. **UserContext.tsx:** This file contains the UserContext and UserProvider components. The UserContext will be used to share user-related data across the application, and the UserProvider will be used to manage the state of this data.
2. **AuthContext.tsx:** This file contains the AuthContext and AuthProvider components. The AuthContext will be used to share authentication-related data (like whether the user is logged in or not) across the application, and the AuthProvider will be used to manage the state of this data.
3. **RequireAuth.tsx:** This file contains your RequireAuth component. This component will be used to protect certain routes that should only be accessible to authenticated users.
4. **api.ts (or similar):** This file contains the functions for making API calls to your backend, like login() and logout(). The AuthProvider will use these functions to authenticate the user.
5. **PrivateRoute.tsx:** This file contains a PrivateRoute component that wraps around RequireAuth. It’s used to create protected routes in a more declarative way with React Router.
6. **LoginForm.tsx:** I used this as a quick login form to manually test login using the Frontend instead of simulating using Postman or ThunderClient extension.

### CODE REVIEW

File: frontend/src/shared/api/types.ts
Comment (glenntanjoh) [inline frontend/src/shared/api/types.ts:1]: Seems this was what I was missing. Nice!

**Review by glenntanjoh — COMMENTED** [review]
Seems you were able to add the files my authentication was missing and also wrapped the routes in the ProtectedRoute hook. Good stuff

**Review by ralphgregorio — APPROVED** [review]
paired and looks good

**Comment by ralphgregorio** [comment]
Leaving the playwright test to fail, need to fix build setup for tests to get fixed.

**Pull request merged**